File: packages/go/model/merkle/proof.go
package merkle

import (
	"encoding/base64"
	"fmt"
	"regexp"
	"strconv"
	"strings"

	utls "github.com/cyrildever/go-utls/common/utils"
	"github.com/cyrildever/merkle-trees/packages/go/exception"
	"github.com/cyrildever/merkle-trees/packages/go/model/hash"
)

// Proof defines the Merkle tree proof that consists of the trail of intermediate hashes suffixed with the path,
// the name of the hashing engine used and the size of the Merkle tree at the date of the proof.
type Proof struct {
	Trail hash.Hashes
	Path
	Size   int
	Engine string
}

// String returns the base64-encoded dot-separated concatenation of the hexadecimal hashes, the path, the engine and the size of the tree, eg.
//
//       (rootHash)
//          /  \
//         ()   hash2
//        /  \
//   (leaf)  hash1
//
//	=> proofStr := Base64Encode("<hash1><hash2>.11.sha-256.4")
func (p *Proof) String() string {
	var hashes []string
	for _, h := range p.Trail {
		hashes = append(hashes, utls.ToHex(h))
	}
	return base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%s.%s.%s.%d", strings.Join(hashes, ""), p.Path, p.Engine, p.Size)))
}

// NewProof ...
func NewProof(trail hash.Hashes, path Path, size int, engine ...string) *Proof {
	usingEngine := hash.SHA_256
	if len(engine) == 1 && engine[0] != "" { // TODO Add engine validity check?
		usingEngine = engine[0]
	}
	return &Proof{
		Trail:  trail,
		Path:   path,
		Size:   size,
		Engine: usingEngine,
	}
}

// ProofFrom builds a `MerkleProof` from the passed string, provided it's an actual stringified proof
func ProofFrom(b64 string) (p *Proof, err error) {
	str, err := base64.StdEncoding.DecodeString(b64)
	if err != nil {
		err = exception.NewInvalidMerkleProofError(b64)
		return
	}
	parts := strings.Split(string(str), ".")
	if len(parts) != 4 {
		err = exception.NewInvalidMerkleProofError(b64)
		return
	}
	var path Path = parts[1]
	engine := parts[2]
	size, err := strconv.Atoi(parts[3])
	if size == 0 || err != nil {
		if err == nil {
			err = exception.NewInvalidMerkleProofError(b64)
		}
		return
	}
	switch engine {
	case hash.SHA_256:
		hashes := regexp.MustCompile("(.{64})").FindAllString(parts[0], -1)
		if hashes == nil {
			err = exception.NewInvalidMerkleProofError(b64)
			return
		}
		var trail hash.Hashes
		for _, h := range hashes {
			bytes, e := utls.FromHex(h)
			if e != nil {
				err = exception.NewInvalidMerkleProofError(b64)
				return
			}
			trail = append(trail, bytes)
		}
		p = NewProof(trail, path, size, engine)
	default:
		err = exception.NewInvalidMerkleProofError(b64)
	}
	return
}

File: packages/py/src/merklepy/proof.py
import base64
from pyutls.hex import from_hex, to_hex
import re

from merklepy.exception import InvalidMerkleProofError
from merklepy.hash import SHA_256, Hashes
from merklepy.path import Path


class Proof(object):
    """
    The proof consists of the trail of intermediate hashes suffixed with the path, the name
    of the hashing engine used and the size of the Merkle tree at the date of the proof.

    Parameters
    ----------
    trail : Hashes
        The intermediate hashes bottom up (not including the final root hash)
    path : Path
        The path from root to leaf
    size : int
        The number of leaves in the Merkle tree at issuance of the proof
    engine : str, optional
        The name of the hashing function (the default is `'sha-256'`)
    """

    def __init__(self, trail: Hashes, path: Path, size: int, engine: str = SHA_256):
        self.trail = trail
        self.path = path
        self.size = size
        self.engine = engine

    def to_string(self) -> str:
        """
        Returns
        -------
        str
            The base64-encoded dot-separated concatenation of the hexadecimal hashes, the path, the engine and the size of the tree, eg.

            Base64('&lt;hash1&gt;&lt;hash2&gt;.11.sha-256.4')
        """
        trail_hex = ''.join(map(to_hex, self.trail))
        readable_proof = '.'.join(
            [trail_hex, self.path, self.engine, str(self.size)])
        return base64.b64encode(bytes(readable_proof, 'utf-8')).decode('utf-8')


def proof_from(string: str) -> Proof:
    """Build a `Proof` from the passed string, provided it's an actual stringified proof"""
    try:
        decoded = str(base64.b64decode(string), 'utf-8')
    except Exception as error:
        raise InvalidMerkleProofError(string) from error
    parts = decoded.split('.')
    if len(parts) != 4:
        raise InvalidMerkleProofError(string)
    path = parts[1]
    engine = parts[2]
    size = int(parts[3])
    if size == 0:
        raise InvalidMerkleProofError(string)
    if engine == SHA_256:
        hashes = list(filter(lambda item: len(item) == 32,
                             map(from_hex, re.split('(.{64})', parts[0]))))
        return Proof(hashes, path, size, engine)
    else:
        raise InvalidMerkleProofError(string)
